packages/firestore/test/unit/util/bundle.test.ts

```diff
@@ -0,0 +1,338 @@
+    }
+  };
+  const doc1String = lengthPrefixedString(doc1);
+
+  const doc2Meta: BundleElement = {
+    documentMetadata: {
+      name:
+        'projects/test-project/databases/(default)/documents/collectionId/doc2',
+      readTime: { seconds: 5, nanos: 6 },
+      exists: true
+    }
+  };
+  const doc2MetaString = lengthPrefixedString(doc2Meta);
+  const doc2: BundleElement = {
+    document: {
+      name:
+        'projects/test-project/databases/(default)/documents/collectionId/doc2',
+      createTime: { seconds: 1, nanos: 2000000 },
+      updateTime: { seconds: 3, nanos: 4000 },
+      fields: { foo: { stringValue: 'value1' }, bar: { integerValue: 42 } }
+    }
+  };
+  const doc2String = lengthPrefixedString(doc2);
+
+  const noDocMeta: BundleElement = {
+    documentMetadata: {
+      name:
+        'projects/test-project/databases/(default)/documents/collectionId/nodoc',
+      readTime: { seconds: 5, nanos: 6 },
+      exists: false
+    }
+  };
+  const noDocMetaString = lengthPrefixedString(noDocMeta);
+
+  const limitQuery: BundleElement = {
+    namedQuery: {
+      name: 'limitQuery',
+      bundledQuery: {
+        parent: 'projects/fireeats-97d5e/databases/(default)/documents',
+        structuredQuery: {
+          from: [{ collectionId: 'node_3.7.5_7Li7XoCjutvNxwD0tpo9' }],
+          orderBy: [{ field: { fieldPath: 'sort' }, direction: 'DESCENDING' }],
+          limit: { 'value': 1 }
+        },
+        limitType: 'FIRST'
+      },
+      readTime: { 'seconds': 1590011379, 'nanos': 191164000 }
+    }
+  };
+  const limitQueryString = lengthPrefixedString(limitQuery);
+  const limitToLastQuery: BundleElement = {
+    namedQuery: {
+      name: 'limitToLastQuery',
+      bundledQuery: {
+        parent: 'projects/fireeats-97d5e/databases/(default)/documents',
+        structuredQuery: {
+          from: [{ collectionId: 'node_3.7.5_7Li7XoCjutvNxwD0tpo9' }],
+          orderBy: [{ field: { fieldPath: 'sort' }, direction: 'ASCENDING' }],
+          limit: { 'value': 1 }
+        },
+        limitType: 'LAST'
+      },
+      readTime: { 'seconds': 1590011379, 'nanos': 543063000 }
+    }
+  };
+  const limitToLastQueryString = lengthPrefixedString(limitToLastQuery);
+
+  async function getAllElement(
+    bundle: BundleReader
+  ): Promise<SizedBundleElement[]> {
+    const result: SizedBundleElement[] = [];
+    while (true) {
+      const sizedElement = await bundle.nextElement();
+      if (sizedElement === null) {
+        break;
+      }
+      if (!sizedElement.isBundleMetadata()) {
+        result.push(sizedElement);
+      }
+    }
+
+    return Promise.resolve(result);
+  }
+
+  function verifySizedElement(
+    element: SizedBundleElement,
+    payload: unknown,
+    payloadString: string
+  ): void {
+    expect(element.payload).to.deep.equal(payload);
+    expect(element.byteLength).to.equal(
+      encoder.encode(payloadString).byteLength
+    );
+  }
+
+  async function parseThroughBundle(
```

Comment:
This reads like "passthrough" but it's not. Maybe `generateBundleAndParse`?

Reply:
Done.